src/number_theory/number_field.lean

```diff
@@ -96,6 +109,19 @@ variables (K)
 
 instance [number_field K] : char_zero (𝓞 K) := char_zero.of_module _ K
 
+/-- The ring of integers of a number field is not a field. -/
+lemma not_is_field [number_field K] : ¬ is_field (𝓞 K) :=
+begin
+  have h_inj: function.injective ⇑(algebra_map ℤ ↥(ring_of_integers K)),
```

Comment:
```suggestion
  have h_inj : function.injective ⇑(algebra_map ℤ ↥(ring_of_integers K)),
```

Reply:
I feel like this should be easier: wouldn't `rw [ring_hom.eq_int_cast]; exact int.cast_injective` work here? If not, could you extract this subproof as a lemma to `data/int/char_zero.lean`?

Reply:
I tried your suggestion and it didn't work. Do you mean extract the following result?
```
lemma algebra_map_injective {α : Type*} [ring α] [algebra ℤ α] [char_zero α] : 
  function.injective ⇑(algebra_map ℤ α) := 
begin
  rw ring_hom.injective_iff,
  intros a ha,
  rw [ring_hom.eq_int_cast, int.cast_eq_zero] at ha,
  exact ha,
end
```
It seems that `data/int/char_zero` doesn't know about `algebra_map`, maybe this could go in `algebra.algebra.basic`.

Reply:
It's more generally true as:
```lean
lemma ring_hom.injective_int {α : Type*} [ring α] (f : ℤ →+* α) [char_zero α] :
  function.injective f :=
subsingleton.elim (int.cast_ring_hom _) f ▸ int.cast_injective
```

Reply:
Thanks! I'll add this to `data.int.char_zero`.

Reply:
```
have h_inj : function.injective ⇑(algebra_map ℤ ↥(ring_of_integers K)),
{ exact ring_hom.injective_int (algebra_map ℤ ↥(ring_of_integers K)) },
```
works fine, but if I insted write it as
```
have h_inj : function.injective ⇑(algebra_map ℤ ↥(ring_of_integers K)) :=
ring_hom.injective_int (algebra_map ℤ ↥(ring_of_integers K)),
```
I get a timeout. Why is this?

Reply:
[This is probably an issue I've run into before](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/infinite.20cached.20failure.20for.20comm_semiring.20S): equality checking of types can be exponentially slow. Because `ring_of_integers K` is a subalgebra, it has a ring structure in multiple ways: via `subsemiring` or via `subring`. And moreover, as a field `K` has a ring structure in multiple ways: via `division_ring` or via `comm_ring`. And all of these have to be unified (repeatedly!) to typecheck `Ζ →+* O K`, which leads to unification problems 50 levels deep and then it times out.

Hopefully there is a way for Lean to remember that these substructures are the same, but for now working around it like changing the elaboration order is the best.

Reply:
Thank you for the explanation!